core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -103,34 +121,45 @@ public void close() throws IOException {
      */
     public void abort() throws IOException {
         close();
-        tmpFile.delete();
+        Files.deleteIfExists(tmpFile);
     }
 
     public void commit() throws IOException {
         close();
-        if (destFile.exists()) {
-            try {
-                Util.deleteFile(destFile);
-            } catch (IOException x) {
-                tmpFile.delete();
-                throw x;
-            }
+        try {
+            // Try to make an atomic move.
+            Files.move(tmpFile, destFile, StandardCopyOption.ATOMIC_MOVE);
+        } catch (AtomicMoveNotSupportedException e) {
+            // If it falls here that means that Atomic move is not supported by the OS.
+            // In this case we need to fall-back to a copy option which is supported by all OSes.
+            Files.move(tmpFile, destFile, StandardCopyOption.REPLACE_EXISTING);
         }
```

Comment:
you are missing catching an `IOException` for the case that `StandardCopyOption.ATOMIC_MOVE` **is** supported but that the Atomic Move can not replace an existing file.

Now this should really be a `FileAlreadyExists` exception but this exception is optional so you really do have to work with a generic `IOException` (yay :-o )